### PR TITLE
Fixed to properly detokenize xml strings containing xml escaped token…

### PR DESCRIPTION
**Pull request by TheJeffer**
Fixed to properly detokenize xml strings containing xml escaped tokens.

Updated TokenParser.ParseString to be simpler, more performant, and safer (now prevents infinite loops in nested parameters).

Added simple unit test for testing nested parameters.

| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| New sample?      | no
| Related issues?  | https://github.com/SharePoint/PnP-Sites-Core/pull/1231

#### What's in this Pull Request?

The TokenParser.ParseString method did not properly handle XML strings containing XML escaped tokens. The token regexes would not match a token that had been XML escaped. I've added a new method that converts the XML string into an XmlDocument, then searches through attributes and element text values, replacing tokens as necessary, then returns back the full XML string.

I've also Updated TokenParser.ParseString to be simpler, more performant, and safer (now prevents infinite loops in nested parameters). Previously, one could cause an infinite loop by using nested params where param A contains a token for param B which contains a token for param C which contains a token for param A. I also added a short-circuiting condition to skip searching a string for tokens if the string doesn't contain a { or ~ character so that we're not needlessly running hundreds of regex searches against a string that doesn't contain any tokens. I don't see a hard requirement for tokens to begin with one of those two characters, but it seems to be the convention for all the tokens currently constructed in this project. If that's not the case, feel free to remove that condition from the initial if statement in the ParseString method.

### CODE REVIEW

File: Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/TokenParser.cs [hunk 1]
```diff
@@ -355,69 +355,57 @@ public IEnumerable<string> GetLeftOverTokens(string input)
...
+                foreach (var node in nodes.OfType<System.Xml.XmlElement>())
+                {
+                    if (!string.IsNullOrEmpty(node.InnerText))
                     {
-                        foreach (var regex in token.GetRegex().Where(regex => regex.IsMatch(input)))
-                        {
-                            origInput = input;
-                            input = regex.Replace(input, token.GetReplaceValue());
-                        }
+                        node.InnerText = ParseString(node.InnerText, tokensToSkip);
```
Comment (SchauDK) [inline Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/TokenParser.cs]: node.InnerText = HttpUtility.HtmlEncode(ParseString(node.InnerText, tokensToSkip));

File: Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/TokenParser.cs [hunk 2]
```diff
@@ -355,69 +355,57 @@ public IEnumerable<string> GetLeftOverTokens(string input)
...
-                            foreach (var filteredToken in filteredTokens)
-                            {
-                                var regex = token.GetRegexForToken(filteredToken);
-                                if (regex.IsMatch(input))
-                                {
-                                    input = regex.Replace(input, token.GetReplaceValue());
-                                }
-                            }
-                        }
+                        attribute.Value = ParseString(attribute.Value, tokensToSkip);
```
Comment (SchauDK) [inline Core/OfficeDevPnP.Core/Framework/Provisioning/ObjectHandlers/TokenParser.cs]: attribute.Value = HttpUtility.HtmlEncode(ParseString(attribute.Value, tokensToSkip));

**Comment by TheJeffer** [comment]
That shouldn't be necessary, since we're using an XmlDocument. The OuterXml property will already give back properly encoded xml. I put together a quick example to show that the values are encoded properly... https://dotnetfiddle.net/1id4dy.

**Comment by SchauDK** [comment]
You're so right. No worries then.
Thanks for the explaination.

**Comment by NicolajHedeager** [comment]
It looks like this change breaks parsing of termset ids for site collection termsets.

You can reproduce this error by adding a site collection term called 'My template set' and the try to parse the string: {termsetid:{sitecollectiontermgroupname}:My template set}

In the previous version of the ParseString method, this would return the id of the termset 'My template set', but the implementation of the method will only parse the {sitecollectiontermgroupname} token and not the termsetid token. This is because the 'while (origInput != input)' loop has been removed.

**Comment by TheJeffer** [comment]
You're absolutely right, @NicolajHedeager. Missed that use case. Fixed the change and added a test for it.

**Comment by jansenbe** [comment]
Hi @TheJeffer, a great change which will be merged. Thanks for your contribution and apologies for the delay in merging.

**Pull request merged**